torch/autograd/function.py

```diff
@@ -288,15 +288,22 @@ def _iter(obj):
 
 
 def _unflatten(input, proto):
+    from torch.nn.utils.rnn import PackedSequence
+
     # unflatten a list or tuple input into a nested list/tuple structure
     # specified by proto
     def unflatten_helper(input, proto):
         res = []
         if not isinstance(proto, (list, tuple)):
             return input[0], input[1:]
         for e in proto:
-            res_e, input = unflatten_helper(input, e)
-            res.append(res_e)
+            if e is None:
+                res.append(e)
+            else:
+                res_e, input = unflatten_helper(input, e)
+                res.append(res_e)
+        if isinstance(proto, PackedSequence):
+            return PackedSequence(*res), input
```

Comment:
Wouldn't changing the code to `type(proto)(*res)` be enough? `isinstance` is very expensive

Reply:
Ah sorry, this will fail for regular tuples, because they expect a single argument in their constructor.

Reply:
Actually where's this even called? I can't see it in the diff

Reply:
@anderspapitto and I agreed that we'll do `type(proto)(res)` and add a constructor to `PackedSequence` which takes a regular tuple.

Reply:
no, because while an n-tuple can be constructed from another n-tuple `x` by saying `tuple(x)`, an n-namedtuple cannot, it must be constructed from n individual arguments.

I'll address this by changing the constructor of PackedSequence() to also accept a single tuple

Reply:
We can change the constructor to always match that of `tuple`. There's a note in the docs that says that it should never be instantiated by the user, but only constructed using the API for packed sequences

Reply:
See my comment above - `namedtuple` has `_make` method which takes the iterable. So you can do `if hasattr(type(proto), '_make')` which is hopefully faster than isinstance

Reply:
I went with changing the constructor to allow either style of arg-passing